### ISSUE TITLE
Data.Set.fromDistinctAscListN

### ISSUE THREAD

**Issue by jwaldmann** [haskell/containers#894]
just making a note that such a function (with N = length of second argument) could be useful, since an implementation can use N to determine the shape of the tree in advance, so there's no need to re-balance during construction.

see https://github.com/haskell/containers/issues/890#issuecomment-1380917641

I made some experiments here https://gitlab.imn.htwk-leipzig.de/waldmann/fdaln and the timing data looks very strange...

**Comment by treeowl** [comment]
Also descending versions?

**Comment by treeowl** [comment]
Your implementations all look weird to me. This is a deserialization problem.

```haskell
fromDistinctAscendingListN :: Int -> [a] -> (Set a, [a])
fromDistinctAscendingListN n xs
  | s :!* remains <- fdalN n xs
  = (s, remains)

data LSP a b = !a :!* b
fdalN :: Int -> [a] -> LSP (Set a) [a]
fdalN n xs | n <= 0 = Tip :!* xs
fdalN 1 (x:xs) = singleton x :!* xs
fdalN n xs
  | start :*! (x : xs') <- fdalN (n `quot` 2) xs
  , end :*! xs'' <- fdalN ((n-1) `quot` 2) xs'
  = Bin n x start end :!* xs''
fdalN _ _ = error "List too short"
```

**Comment by treeowl** [comment]
Mine looks a lot like a direct version of your continuation-passing ones. I'd expect it to be easier on the optimizer.

**Comment by treeowl** [comment]
I agree none of this makes sense. There's no reasonable world in which adding the overhead of `STRef` and such should improve matters. Is GHC doing something very silly? Was your system reasonably quiet? Were you getting pretty consistent results?

**Comment by treeowl** [comment]
I do think your test is a bit unusual. First off, why sum the sets at all? Why not just force them to WHNF? They're totally strict structures. Secondly, why not use a common benchmark system? `containers` uses `tasty-bench` because it doesn't have a lot of dependencies. How did you choose to use a list in memory rather than one that's generated? Does that make a difference to the relative performance?

**Comment by meooow25** [comment]
I haven't checked the linked implementations but noticed this issue so I thought I should mention, the amount of rebalancing work in `Set.fromDistinctAscList` is just $O(\log n)$. Which is admittedly not zero, but very little compared to the total $O(n)$ time. So `fromDistinctAscListN` vs `fromDistinctAscList` will come down to constant factor, and the benefit if any will likely be very small.

(See also #949)
